features/FEATURE_BLE/source/generic/GenericSecurityManager.cpp

```diff
@@ -0,0 +1,1240 @@
+        } else {
+            return slave_security_request(connection);
+        }
+    }
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::setHintFutureRoleReversal(bool enable) {
+    _master_sends_keys = enable;
+    return BLE_ERROR_NONE;
+}
+
+////////////////////////////////////////////////////////////////////////////
+// Encryption
+//
+
+ble_error_t GenericSecurityManager::getLinkEncryption(
+    connection_handle_t connection,
+    link_encryption_t *encryption
+) {
+
+    ControlBlock_t *cb = get_control_block(connection);
+    if (!cb) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    if (cb->encrypted) {
+        if (cb->ltk_mitm_protected  || cb->mitm_performed) {
+            *encryption = link_encryption_t::ENCRYPTED_WITH_MITM;
+        } else {
+            *encryption = link_encryption_t::ENCRYPTED;
+        }
+    } else if (cb->encryption_requested) {
+        *encryption = link_encryption_t::ENCRYPTION_IN_PROGRESS;
+    } else {
+        *encryption = link_encryption_t::NOT_ENCRYPTED;
+    }
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::setLinkEncryption(
+    connection_handle_t connection,
+    link_encryption_t encryption
+) {
+    ControlBlock_t *cb = get_control_block(connection);
+    if (!cb) {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    link_encryption_t current_encryption(link_encryption_t::NOT_ENCRYPTED);
+    getLinkEncryption(connection, &current_encryption);
+
+    if (current_encryption == link_encryption_t::ENCRYPTION_IN_PROGRESS) {
+        return BLE_ERROR_OPERATION_NOT_PERMITTED;
+    }
+
+    /* ignore if the link is already at required state*/
+    if (current_encryption == encryption) {
+        return BLE_ERROR_NONE;
+    }
+
+    if (encryption == link_encryption_t::NOT_ENCRYPTED) {
+
+        if (cb->encrypted) {
+            return _pal.disable_encryption(connection);
+        }
+
+    } else if (encryption == link_encryption_t::ENCRYPTED) {
+
+        /* if already better than encrypted don't bother */
+        if (current_encryption == link_encryption_t::ENCRYPTED_WITH_MITM) {
+            return BLE_ERROR_NONE;
+        }
+        cb->encryption_requested = true;
+        return enable_encryption(connection);
+
+    } else if (encryption == link_encryption_t::ENCRYPTED_WITH_MITM) {
+
+        if (cb->ltk_mitm_protected && !cb->encrypted) {
+            cb->encryption_requested = true;
+            return enable_encryption(connection);
+        } else {
+            cb->encryption_requested = true;
+            return requestAuthentication(connection);
+        }
+
+    } else {
+        return BLE_ERROR_INVALID_PARAM;
+    }
+
+    return BLE_ERROR_NONE;
+}
+
+ble_error_t GenericSecurityManager::getEncryptionKeySize(
+    connection_handle_t connection,
+    uint8_t *size
+) {
+    ControlBlock_t *cb = get_control_block(connection);
+    if (cb) {
+        *size = cb->encryption_key_size;
```

Comment:
It looks like the encryption_key_size is set nowhere.

Reply:
we don't get that information yet

we would need it at pairing successfully result